daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/ScenarioRunner.scala

```diff
@@ -436,8 +443,15 @@ object ScenarioRunner {
                 case Right(_) => go()
               }
             case Some(coinst) => {
-              callback(coinst)
-              go()
+              if (!ledger.isActive(coid)) {
```

Comment:
I think this is the wrong place to check this. We should follow how a real ledger, e.g., SoX works:

1. During submission we accept all disclosed contracts. We don’t care if they’re active or not.
2. Then during validation we check that they’re actually active. In our case that’s in `ScenarioLedger.commitTransaction`.

Reply:
but the activeness check wouldn't be possible for the party that gets a disclosed contract on a real ledger, because it doesn't know its history?

Reply:
> but the activeness check wouldn't be possible for the party that gets a disclosed contract on a real ledger, because it doesn't know its history?

The stakeholders of that contract will validate the transaction and those can check activeness. There is no difference between divulged contracts which we already have & disclosed contracts. In both cases you rely on stakeholders to check activeness.

Reply:
Let’s look at what a "real ledger" (e.g. sandbox on x) does:

A: First without disclosed contracts:

1. During submission, all lookups are resolved against committed ledger state known to the participant. So ResultNeedContract checks that the contract exists and has not been archived _at this point_.
2. Once the participant is happy, it submits the transaction to the network. Then you have some component that sequences transactions (canton sequencer, vmbc committer, sandbox on x bridge) and you check that the contract is still active at the point it was sequenced.

B: Now with disclosed contracts:
1. During submissions, disclosed contracts are always considered active. The participant in general doesn’t even have those contracts in its committed ledger state so there is no way you can do an activeness check here.
2. After sequencing, you check that all contracts used in the transaction are still active. For centralized architectures like sox & vmbc you can check that for all contracts. For Canton, each participant only checks it for stakeholders hosted on that participant. Note that at point there is no difference between disclosed contracts or not.

Now let’s look at the scenario service:
A.1: This happens in `lookupContract` as you correctly pointed out.
A.2: This check does not happen at all. The scenario runner doesn’t have concurrent submissions so the ledger state doesn’t change between when we interpreted the transaction and when it’s committed so we can skip that.

So now for disclosed contracts: In general, I think we should aim to keep the overall execution of the scenario service close to "real ledgers" and not diverge unless we have a strong reason for doing so. Given that, we get this:

B.1: Do the same what other ledgers do: always treat those contracts as active. Once you move those contracts into speedy this becomes even easier since you’re using the exact same logic for both the scenario service and anything else for looking up those contracts.
B.2: Add the check in commitTransaction that contracts are really active. You can either try to be clever and only do that check for disclosed contracts or do it for everything. I’d probably try to do it only for disclosed contracts but I haven’t checked how much complexity that adds.

If you’re looking for some prerequisites of this type of check consider duplicate contract key checks: If I submit a single create, Daml Engine does not check that there is no other contract with that key. That only gets checked when the transaction is committed. That’s why `committTransaction` includes this check.